docs/05-engineering/git-gitlab-github-bitbucket.md

```diff
@@ -1,29 +1,144 @@
+  - If you need to make a hotfix make a PR/MR to the main `master` branch first before creating/patching a hotfix branch.
+- Feature branches undergo both automatic and manual testing before being merged - this ensures that master always passes tests.
+  - Testing often includes linting, unit and functional tests as well as specialised tests such as security and accessibility scans.
+  - On most projects a dedicated review environment for each feature branch or PR/MR is created automatically and can be used for manual testing as well as stakeholder demos and user testing as needed to validate functionality before the branch is merged.
+
+Compared to other documented workflows:
+
+- [https://docs.gitlab.com/ee/topics/gitlab_flow.html](Gitlab Flow) - very similar, although we prefer tags over environment branches
+- [https://guides.github.com/introduction/flow/](GitHub Flow) - very similar, although we prefer rebase/fast-foward merges and don't deploy feature branches into production (but to review environments instead)
+- [Git flow](https://nvie.com/posts/a-successful-git-branching-model/) - not very similar - ours is much simpler, avoids long lived branches and repeated merges
+
+## Versioning releases
+
+_Note: your project may include additional or alternate versioning practices - please review your project developer/contributor documentation!_
+
+For libraries, we use [semantic versioning](https://semver.org/).
+
+For web sites or applications we use an `vX.Y.Z` syntax with the following meaning:
+
+- X: The sprint number
+- Y: A zero-based counter indicating the number of releases to the stage environment
+- Z: A zero-based counter indicating the number of hotfix releases to the prod environment
+
+So in the normal release flow:
+
+- First sprint, pushes to stage has tag v1.0.0
+- Further work and a v1.1.0 tag is created and pushed to stage
+- Over the course of the sprint you might get v1.2.0, v1.3.0 etc
+- Sprint 1 ends, v1.3.0 looks good, gets pushed to prod
+
+In the case a subsiquent hotfix is required:
+
+- In Sprint 2 work continues on master, v2.0.0 tag gets pushed to stage etc.
+- Meanwhile, critical bugfix needed on prod, so the v1.3.0 is branched to v1.
+  - There is no need for further suffixes in the branch name, since it's the only branch for sprint 1
+  - We don't typically create these branches until they are needed, since it's hopefully rare to need critical hotfixes and we have a tag documenting the state of the release.
+- Fix is pushed to master (i.e. upstream-first workflow), then cherry-picked to v1.
+- Tag v1.3.1 is then created from the v1 branch, pushed to stage (or some dedicated hotfix stage environment) and then to prod.
+- If another hotfix is needed, that commit would also be merged to master, cherry-picked into v1 and tagged v1.3.2 etc.
+
+In a continuous deployment project these tags can be created automatically, or some other mechanism may be used to track releases.
```

Comment:
I have not used this. It sounds interesting. Will have to try and practice it, or thing about it some more :)